ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Types.hs

```diff
@@ -635,6 +638,47 @@ data AbstractState
     | TerminatedSt
     deriving (Eq, Show, Typeable)
 
+-- | Counters for tracing and analysis purposes
+--
+data ConnectionManagerCounters = ConnectionManagerCounters {
+  numberConns   :: !Int, -- ^ number of connections relevant for pruning
+  duplexConns   :: !Int, -- ^ number of negotiated duplex connections (including DuplexState connections)
+  uniConns      :: !Int, -- ^ number of negotiated unidirectional connections
+  incomingConns :: !Int, -- ^ number of inbound connections
+  outgoingConns :: !Int  -- ^ number of outbound connections
+  } deriving Show
+
+instance Semigroup ConnectionManagerCounters where
+    ConnectionManagerCounters c1 d1 s1 i1 o1 <> ConnectionManagerCounters c2 d2 s2 i2 o2 =
+      ConnectionManagerCounters (c1 + c2) (d1 + d2) (s1 + s2) (i1 + i2) (o1 + o2)
+
+instance Monoid ConnectionManagerCounters where
+    mempty = ConnectionManagerCounters 0 0 0 0 0
+
+-- | Perform counting from an 'AbstractState'
+abstractStateToCounters :: AbstractState -> ConnectionManagerCounters
+abstractStateToCounters state =
```

Comment:
I'd suggest to use `ConnectionState` not `AbstractState`: this introduces unnecessary indirection. `AbstractState` is here only for nicer tracing messages, but it really belongs to the testing framework.

Reply:
Alright, will do! 😄